kyuubi-server/src/main/scala/org/apache/kyuubi/client/KyuubiSyncThriftClient.scala

```diff
@@ -109,6 +110,18 @@ class KyuubiSyncThriftClient private (
           }
         } else {
           shutdownAsyncRequestExecutor()
+          Option(engineAliveThreadPool).foreach { pool =>
+            ThreadUtils.shutdown(pool, Duration(engineAliveProbeInterval, TimeUnit.MILLISECONDS))
+          }
```

Comment:
self-terminating? looks like not feasible

Reply:
Thank you for the comment.
May I ask for more details, please?

Reply:
How about simply removing it? cc @ulysses-you

Reply:
It would cause leak. Move below the code `clientClosedOnEngineBroken = true` should be fine.

Reply:
Thank you for your reply here.

May I know why the code block has to be moved under the `clientClosedOnEngineBroken = true` please?

In the KyuubiSyncThriftClient#closeSession
`ThreadUtils.shutdown(pool, Duration(engineAliveProbeInterval, TimeUnit.MILLISECONDS))` comes before the `tProtocol.getTransport.close()`

https://github.com/apache/kyuubi/blob/8267108e7f4ff282aa0f56639a81a1bdf992ff6d/kyuubi-server/src/main/scala/org/apache/kyuubi/client/KyuubiSyncThriftClient.scala#L212-L226

Thank you for your patience.

Reply:
In `KyuubiSyncThriftClient#closeSession`, we shutdown probe thread in another thread, but now we shutdown probe thread by itself. It should be more safer to make sure we have closed all things before shutdown itself.

Reply:
Thank you, now I fully understand your previous comment.

I have shifted the following code after the `clientClosedOnEngineBroken = true`, to make sure that no code has to be executed after closing the thread.
```
Option(engineAliveThreadPool).foreach { pool => 
   ThreadUtils.shutdown(pool, Duration(engineAliveProbeInterval, TimeUnit.MILLISECONDS)) 
 } 
```

By the way, re-confirmed that "self-terminating" is working.
```scala
import java.util.concurrent._

object ThreadTest {
  def main(args: Array[String]) = {
    val scheduledExecutorService: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
    val myTask: Runnable = new Runnable {
      def run(): Unit = {
        if(true) {
          println("Thread shutdown start")
          scheduledExecutorService.shutdownNow()
          println("Thread shutdown finish")
        }
      }
    }
    scheduledExecutorService.scheduleWithFixedDelay(myTask, 1, 1, TimeUnit.SECONDS)
    while (!scheduledExecutorService.isShutdown()) {
      Thread.sleep(400);
      println("ScheduledExecutorService is still running")
    }
    println("ScheduledExecutorService is shut down")
  }
}
```
```bash
### Output ###
$ scala ThreadTest
ScheduledExecutorService is still running
ScheduledExecutorService is still running
Thread shutdown start
Thread shutdown finish
ScheduledExecutorService is still running
ScheduledExecutorService is shut down
```